security/tools/generate_cert/main.go

```diff
@@ -19,66 +19,96 @@ package main
 import (
 	"crypto"
 	"crypto/x509"
+	"encoding/json"
 	"flag"
 	"fmt"
 	"io/ioutil"
-	"os"
+	"log"
+	"os/exec"
 	"time"
 
-	"istio.io/istio/pkg/log"
 	"istio.io/istio/security/pkg/pki/util"
+	k8s "k8s.io/api/core/v1"
 )
 
-// Layout for parsing time
-const timeLayout = "Jan 2 15:04:05 2006"
+const (
+	// Layout for parsing time
+	timeLayout = "Jan 2 15:04:05 2006"
+
+	// modes supported by this tool.
+	selfSignedMode = "self-signed"
+	signerMode     = "signer"
+	citadelMode    = "citadel"
+)
 
 var (
 	host           = flag.String("host", "", "Comma-separated hostnames and IPs to generate a certificate for.")
 	validFrom      = flag.String("start-date", "", "Creation date in format of "+timeLayout)
 	validFor       = flag.Duration("duration", 365*24*time.Hour, "Duration that certificate is valid for.")
 	isCA           = flag.Bool("ca", false, "Whether this cert should be a Cerificate Authority.")
-	isSelfSigned   = flag.Bool("self-signed", false, "Whether this cerificate is self-signed.")
 	signerCertFile = flag.String("signer-cert", "", "Signer certificate file (PEM encoded).")
 	signerPrivFile = flag.String("signer-priv", "", "Signer private key file (PEM encoded).")
 	isClient       = flag.Bool("client", false, "Whether this certificate is for a client.")
 	org            = flag.String("organization", "Juju org", "Organization for the cert.")
 	outCert        = flag.String("out-cert", "cert.pem", "Output certificate file.")
 	outPriv        = flag.String("out-priv", "priv.pem", "Output private key file.")
 	keySize        = flag.Int("key-size", 2048, "Size of the generated private key")
+	mode           = flag.String("mode", selfSignedMode, "Supported mode: self-signed, signer, citadel")
 )
 
-func fatalf(template string, args ...interface{}) {
-	log.Errorf(template, args)
-	os.Exit(-1)
-}
-
 func checkCmdLine() {
 	flag.Parse()
 
 	hasCert, hasPriv := len(*signerCertFile) != 0, len(*signerPrivFile) != 0
-	if *isSelfSigned {
+	switch *mode {
+	case selfSignedMode:
 		if hasCert || hasPriv {
-			fatalf("--self-signed is incompatible with --signer-cert or --signer-priv.")
+			log.Fatal("--self-signed is incompatible with --signer-cert or --signer-priv.")
 		}
-	} else {
-		if !hasCert && !hasPriv {
-			fatalf("Need --self-signed or --signer-cert and --signer-priv.")
-		} else if !(hasCert && hasPriv) {
-			fatalf("Missing --signer-cert or --signer-priv.")
+	case signerMode:
+		if !hasCert || !hasPriv {
+			log.Fatalf("Need --self-signed or --signer-cert and --signer-priv.")
```

Comment:
this error msg is a bit confusing, probably remove ```--self-signed```  ?

Reply:
Done.